prot.c

```diff
@@ -9,7 +9,19 @@ int prot_gid(gid_t gid)
   return setgid(gid); /* _should_ be redundant, but on some systems it isn't */
 }
 
-int prot_uid(uid_t uid)
+#ifndef GROUP_COUNT
+/* use a sane default, but one can override it from outside if there are
+ * higher requirements. */
+#define GROUP_COUNT 8
+#endif
+int prot_gids(const char *user, gid_t gid)
 {
-  return setuid(uid);
+  gid_t gids[GROUP_COUNT];
+  int gcount = GROUP_COUNT;
+  int r = getgrouplist(user, gid, gids, &gcount);
+  /* member of too many groups */
+  if (r < 0)
+    return r;
+  if (setgroups(gcount,gids) == -1) return -1;
+  return setgid(gid); /* _should_ be redundant, but on some systems it isn't */
 }
```

Comment:
Untested, but how about this attempted refactoring of `prot.h` and `prot.c`:

```c
#ifndef PROT_H
#define PROT_H

#include <sys/types.h>
#include <unistd.h>

#define prot_gid(gid) prot_gids(NULL, gid);
extern int prot_gids(const char *user, gid_t gid);
#define prot_uid(uid) setuid(uid)

#endif
```

```c
#include <sys/types.h>
#include <unistd.h>
#include <grp.h>
#include "prot.h"

#ifndef GROUP_COUNT
/* use a sane default, but one can override it from outside if there are
 * higher requirements. */
#define GROUP_COUNT 8
#endif
int prot_gids(const char *user, gid_t gid)
{
  gid_t gids[GROUP_COUNT];
  int gcount = GROUP_COUNT;

  if (user) {
    int r = getgrouplist(user, gid, gids, &gcount);
    /* member of too many groups */
    if (r < 0)
      return r;
  } else {
    gcount = 1;
    gids[0] = gid;
  }

  if (setgroups(gcount,gids) == -1) return -1;
  return setgid(gid); /* _should_ be redundant, but on some systems it isn't */
}
```

Reply:
Lovely.